### ISSUE TITLE
Settings to stop beta crashing / UI not drawing/updating?

### ISSUE THREAD

**Issue by johnmeat** [magefree/mage#8968]
I've recently downloaded the beta, nearly all games against AI end by freezing where the UI doesn't draw properly and the game is frozen, are there any settings I can change to help this? This is both connecting to the beta server and running locally thanks

![image](https://user-images.githubusercontent.com/64497504/168991875-a5ea4046-3266-4720-86f9-5b476b2e0672.png)

**Comment by JayDi85** [comment]
More details and fixes example in the #5023

**Comment by johnmeat** [comment]
Thanks I've changed the AI decks and it appears to be running much better.

**Comment by Alex-Vasile** [comment]
> Thanks I've changed the AI decks and it appears to be running much better.

Which cards were the issue?

**Comment by johnmeat** [comment]
I haven't looked into it. I was playing against the Edgar Markov precon and a Scarab God edhrec average deck, and just changed those for other decks. When hanging it wasn't ever particular an effect or card trying to resolve. For instance once it was on a draw or upkeep of an AI with nothing on the stack.

**Comment by Alex-Vasile** [comment]
> I haven't looked into it. I was playing against the Edgar Markov precon and a Scarab God edhrec average deck, and just changed those for other decks. When hanging it wasn't ever particular an effect or card trying to resolve. For instance once it was on a draw or upkeep of an AI with nothing on the stack.

Thanks for the info!

What were you playing?

**Comment by johnmeat** [comment]
[[Errant, Street Artist]]

**Comment by jeffwadsworth** [comment]
Perhaps we should set the AI level at 1.  In my experience, it plays just fine unless you are using a Texas Instruments calculator circa 1975, etc.

**Comment by jeffwadsworth** [comment]
There are some pretty sweet tests in the suite for this already, I believe.  They are just disabled.

**Comment by JayDi85** [comment]
Yes, I'm using `LoadTest` for testing new sets and AI stability -- it allows to play random cards by AI as fast as possible and to find errors or freezes. Game logs can help to find a buggy generated deck and repeat it in next run by random seed.

**Comment by Alex-Vasile** [comment]
> Yes, I'm using `LoadTest` for testing new sets and AI stability -- it allows to play random cards by AI as fast as possible and to find errors or freezes. Game logs can help to find a buggy generated deck and repeat it in next run by random seed.

I'm looking through that and it seems to be only randomly generated decks. Is there functionality already built for loading up specific decks from a .dck file?

**Comment by Alex-Vasile** [comment]
Alright, looking into this further a) the CG changes I was trying does no help performance.

But b), a big part of the issue appears to be a duplicate of #2218.

Both the Edgar Marvok and Scarab God decks create a lot of tokens as well as contain lots of targeted spells and abilities (targeted removal, reanimation, and equip).

E.g.
```
INFO  2022-05-31 09:08:13,456 -> #4 (Cast Drown in the Loch) 
INFO  2022-05-31 09:08:13,456 -> #5 (Cast Drown in the Loch) 
INFO  2022-05-31 09:08:13,456 -> #6 (Cast Drown in the Loch) 
INFO  2022-05-31 09:08:13,456 -> #7 (Cast Drown in the Loch) 
INFO  2022-05-31 09:08:13,456 -> #8 (Cast Drown in the Loch) 
INFO  2022-05-31 09:08:13,456 -> #9 (Cast Drown in the Loch) 
INFO  2022-05-31 09:08:13,456 -> #10 (Cast Drown in the Loch)
INFO  2022-05-31 09:08:13,456 -> #11 (Cast Drown in the Loch)
INFO  2022-05-31 09:08:13,456 -> #12 (Cast Drown in the Loch)
INFO  2022-05-31 09:08:13,456 -> #13 (Cast Drown in the Loch)
INFO  2022-05-31 09:08:13,456 -> #14 (Cast Drown in the Loch)
INFO  2022-05-31 09:08:13,456 -> #15 (Cast Drown in the Loch)

INFO  2022-05-31 09:08:13,456 -> #16 ({2}{U}{B}: Exile target creature card from a graveyard. Create a token that's a copy of it, except it's a 4/4 black Zombie.)
INFO  2022-05-31 09:08:13,456 -> #17 ({2}{U}{B}: Exile target creature card from a graveyard. Create a token that's a copy of it, except it's a 4/4 black Zombie.)
INFO  2022-05-31 09:08:13,456 -> #18 ({2}{U}{B}: Exile target creature card from a graveyard. Create a token that's a copy of it, except it's a 4/4 black Zombie.)
INFO  2022-05-31 09:08:13,456 -> #19 ({2}{U}{B}: Exile target creature card from a graveyard. Create a token that's a copy of it, except it's a 4/4 black Zombie.)
INFO  2022-05-31 09:08:13,456 -> #20 ({2}{U}{B}: Exile target creature card from a graveyard. Create a token that's a copy of it, except it's a 4/4 black Zombie.)
INFO  2022-05-31 09:08:13,456 -> #21 ({2}{U}{B}: Exile target creature card from a graveyard. Create a token that's a copy of it, except it's a 4/4 black Zombie.)
INFO  2022-05-31 09:08:13,456 -> #22 ({2}{U}{B}: Exile target creature card from a graveyard. Create a token that's a copy of it, except it's a 4/4 black Zombie.)
INFO  2022-05-31 09:08:13,456 -> #23 ({2}{U}{B}: Exile target creature card from a graveyard. Create a token that's a copy of it, except it's a 4/4 black Zombie.)
INFO  2022-05-31 09:08:13,456 -> #24 ({2}{U}{B}: Exile target creature card from a graveyard. Create a token that's a copy of it, except it's a 4/4 black Zombie.)
INFO  2022-05-31 09:08:13,456 -> #25 ({2}{U}{B}: Exile target creature card from a graveyard. Create a token that's a copy of it, except it's a 4/4 black Zombie.)
INFO  2022-05-31 09:08:13,456 -> #26 ({2}{U}{B}: Exile target creature card from a graveyard. Create a token that's a copy of it, except it's a 4/4 black Zombie.)
INFO  2022-05-31 09:08:13,456 -> #27 ({2}{U}{B}: Exile target creature card from a graveyard. Create a token that's a copy of it, except it's a 4/4 black Zombie.)
```

The AI never actually freezes, what's going on is that the combination of possible options becomes incredibly large, especially when the AI difficulty is set much beyond 1.

For the attached piece of the log, I am not sure what to do about speeding up the activated ability since there are very few repeats of those targets.

But for [[Drown in the Loch]], many of those instances are targeting identical tokens (stack was empty so AI is using destroy part of spell). Pruning identical targets from the list should provide a speedup. It also does not make sense for the AI to consider each identical token as a separate target since we know ahead of time that the result would be the same regardless of which is chosen.

**Comment by Alex-Vasile** [comment]
The problem was related to Edgar Markov. Closed by 420b6f1a74590b5126a5525fdb86d8e4f70916dd.

**Comment by Alex-Vasile** [comment]
> Some additional info about AI logic/code:
> 
> 1. AI play's spell and targets uses only unique names, so it will see same ability/target as one after optimization;
> 2. Targeting for targeted spell and targeting as a result of some spell/ability are different for AI (the first one uses game simulations with all possible targets, the second uses same best targets due target logic).
> 3. AI level equal to "add additional seconds" to the think time. After 20-30 seconds of game simulations -- AI will force to stop all code and choose the best one.

Where does Step 1 happen? I haven't found a portion of the code that does this.
